### PR TITLE
End2End Video SFT with Varied Size/Duration

### PR DESCRIPTION
**Pull request by hengtaoguo**
# Description

- Add `sft-vision-llava-video-178k.yml` for end-to-end video SFT support for Qwen3*, supporting variable video sizes and durations.
- Extend the Hugging Face input pipeline with ShareGPT formatting, video detection, decoding, preprocessing, padding, and GCS support.
- Add LLaVA-Video-178K preparation tooling and an example training configuration (MVP loading video locally).
- Route video tensors, masks, and grid metadata through training while preserving existing image SFT behavior.
- AI-generated documentation for this change: [link](https://paste.googleplex.com/5423038980554752).

# Tests

Dataset download/unzip to local machine and parquet generation:
```
python tools/data_generation/prepare_llava_video_178k.py --output_dir /mounted/LLaVA-Video-178K --fold 0_30_s_academic_v0_1 --download_videos
```

10-step SFT loss decreased from 1.360 to 1.106:
```
# Command
python -m maxtext.trainers.post_train.sft.train_sft_native src/maxtext/configs/post_train/sft-vision-llava-video-178k.yml model_name=qwen3-vl-2b tokenizer_path=Qwen/Qwen3-VL-2B-Instruct base_output_directory=/home/hengtaoguo_google_com/projects/sft_output scan_layers=false per_device_batch_size=2 dataset_type=hf steps=10 video_max_grid_t=32 video_max_grid_h=32 video_max_grid_w=32 max_prefill_predict_length=8192 max_target_length=9216 checkpoint_period=100 use_audio_in_video=False attention=dot_product attention_for_vit=autoselected enable_checkpointing=true hf_train_files=gs://hengtaoguo-maxtext-logs/datasets/LLaVA-Video-178K/0_30_s_academic_v0_1/\*.parquet

# Training logs from scratch
I0805 21:54:24.075999 140641926052992 metric_logger.py:226] completed step: 1, seconds: 12.405, TFLOP/s/device: 20.049, Tokens/s/device: 1485.815, total_weights: 2004, loss: 1.360, lm_loss: 1.360, perplexity: 3.897
I0805 21:54:27.786081 140641926052992 metric_logger.py:226] completed step: 2, seconds: 0.997, TFLOP/s/device: 249.343, Tokens/s/device: 18478.585, total_weights: 2003, loss: 1.240, lm_loss: 1.240, perplexity: 3.456
I0805 21:54:34.939766 140641926052992 metric_logger.py:226] completed step: 3, seconds: 4.145, TFLOP/s/device: 59.996, Tokens/s/device: 4446.292, total_weights: 1899, loss: 1.200, lm_loss: 1.200, perplexity: 3.322
I0805 21:54:43.938847 140641926052992 metric_logger.py:226] completed step: 4, seconds: 9.370, TFLOP/s/device: 26.544, Tokens/s/device: 1967.149, total_weights: 2018, loss: 1.195, lm_loss: 1.195, perplexity: 3.303
I0805 21:54:54.417748 140641926052992 metric_logger.py:226] completed step: 5, seconds: 8.999, TFLOP/s/device: 27.638, Tokens/s/device: 2048.213, total_weights: 1938, loss: 1.051, lm_loss: 1.051, perplexity: 2.860
I0805 21:55:03.992380 140641926052992 metric_logger.py:226] completed step: 6, seconds: 10.479, TFLOP/s/device: 23.735, Tokens/s/device: 1758.958, total_weights: 1565, loss: 1.221, lm_loss: 1.221, perplexity: 3.391
I0805 21:55:14.732353 140641926052992 metric_logger.py:226] completed step: 7, seconds: 9.575, TFLOP/s/device: 25.976, Tokens/s/device: 1925.083, total_weights: 2066, loss: 1.133, lm_loss: 1.133, perplexity: 3.106
I0805 21:55:24.081050 140641926052992 metric_logger.py:226] completed step: 8, seconds: 10.740, TFLOP/s/device: 23.158, Tokens/s/device: 1716.224, total_weights: 1728, loss: 1.106, lm_loss: 1.106, perplexity: 3.021
```

SFT checkpoint decode (describing the dinosaur slightly more accurate without `multiple heads`).
```
# Decode with SFT checkpoint
python -m maxtext.inference.decode maxtext/configs/base.yml tokenizer_type=huggingface model_name=qwen3-vl-2b tokenizer_path=Qwen/Qwen3-VL-2B-Instruct load_parameters_path=/mounted/sft_output/ht_test/checkpoints/9/items per_device_batch_size=1 run_name=ht_test scan_layers=false use_multimodal=true prompt=\'Describe\ what\ you\ can\ see\ in\ this\ video.\' video_path=\'tests/assets/test_video.mp4\' max_prefill_predict_length=1240 max_target_length=1280 ici_tensor_parallelism=4 override_model_config=true attention=\'dot_product\' hf_access_token=xxx

# Before
<|vision_start|><|video_pad|><|vision_end|>Describe what you can see in this video.<|im_end|>
<|im_start|>assistant
` -> `The video shows a large, detailed model of a dinosaur, likely a Tyrannosaurus rex, with multiple heads. The dinosaur is positioned in a dynamic pose, with its body angled forward and its head`

# After
Input `<|im_start|>user
<|vision_start|><|video_pad|><|vision_end|>Describe what you can see in this video.<|im_end|>
<|im_start|>assistant
` -> `The video begins with a close-up of a large, brown dinosaur model with a textured skin pattern, standing amidst green ferns. The dinosaur's mouth is open, revealing sharp teeth and a pink interior,`
```

Run image-sft to make sure no regression introduced (thanks to @subawocit 's help). Last step's loss is identical ([full training logs for both runs](https://paste.googleplex.com/4525084898361344)).
```
# 20-step SFT with coco dataset on main and current branch
python -m maxtext.trainers.post_train.sft.train_sft_native src/maxtext/configs/post_train/sft-vision-coco-captions.yml model_name=qwen3-vl-2b load_parameters_path=/home/hengtaoguo_google_com/projects/checkpoints/qwen3-vl-2b/unscanned/2026-07-31-10-32/0/items base_output_directory=/dev/shm/hengtaoguo/sft_output run_name=main ici_tensor_parallelism=4 per_device_batch_size=1 max_prefill_predict_length=1024 max_target_length=2048 grain_worker_count=0 steps=20 remat_policy=full scan_layers=false save_checkpoint_on_start=false save_checkpoint_on_completion=true
python -m maxtext.trainers.post_train.sft.train_sft_native src/maxtext/configs/post_train/sft-vision-coco-captions.yml model_name=qwen3-vl-2b load_parameters_path=/home/hengtaoguo_google_com/projects/checkpoints/qwen3-vl-2b/unscanned/2026-07-31-10-32/0/items base_output_directory=/dev/shm/hengtaoguo/sft_output run_name=branch ici_tensor_parallelism=4 per_device_batch_size=1 max_prefill_predict_length=1024 max_target_length=2048 grain_worker_count=0 steps=20 remat_policy=full scan_layers=false save_checkpoint_on_start=false save_checkpoint_on_completion=true

# Decode respectively
python -m maxtext.inference.decode maxtext/configs/base.yml tokenizer_type=huggingface model_name=qwen3-vl-2b tokenizer_path=Qwen/Qwen3-VL-2B-Instruct load_parameters_path=/dev/shm/hengtaoguo/sft_output/main/checkpoints/19/items per_device_batch_size=1 run_name=ht_test scan_layers=false use_multimodal=true prompt=\'Describe\ what\ you\ can\ see\ in\ this\ video.\' video_path=\'tests/assets/test_video.mp4\' max_prefill_predict_length=1240 max_target_length=1340 ici_tensor_parallelism=4 override_model_config=true attention=\'dot_product\' hf_access_token=xxx
python -m maxtext.inference.decode maxtext/configs/base.yml tokenizer_type=huggingface model_name=qwen3-vl-2b tokenizer_path=Qwen/Qwen3-VL-2B-Instruct load_parameters_path=/dev/shm/hengtaoguo/sft_output/branch/checkpoints/19/items per_device_batch_size=1 run_name=ht_test scan_layers=false use_multimodal=true prompt=\'Describe\ what\ you\ can\ see\ in\ this\ video.\' video_path=\'tests/assets/test_video.mp4\' max_prefill_predict_length=1240 max_target_length=1340 ici_tensor_parallelism=4 override_model_config=true attention=\'dot_product\' hf_access_token=xxx

# Results identical
Input `<|im_start|>user
<|vision_start|><|video_pad|><|vision_end|>Describe what you can see in this video.<|im_end|>
<|im_start|>assistant
` -> `A large brown dinosaur standing in the middle of green plants.
```

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/multimodal/processor_qwen3_omni.py [hunk 1]
```diff
@@ -471,25 +474,29 @@ def _read_video_decord(video_path, video_start=0.0, video_end=None) -> tuple[np.
...
-  sample_fps = nframes / max(total_frames, 1e-6) * video_fps
-  return video, sample_fps
+    sample_fps = nframes / max(total_frames, 1e-6) * video_fps
+    return video, sample_fps
+  except Exception as e:  # pylint: disable=broad-exception-caught
+    logging.warning("Failed to read/decode video %s: %s. Using dummy video.", video_path, e)
+    # Return dummy video: 4 frames of 224x224 black pixels
+    # 224 is a multiple of 28 (patch_size * merge_size = 14 * 2 = 28)
+    dummy_video = np.zeros((4, 3, 224, 224), dtype=np.uint8)
+    return dummy_video, 2.0
```
Comment (hengtaoguo) [inline src/maxtext/multimodal/processor_qwen3_omni.py:499]: Some videos in dataset are corrupted so this will fallback to dummy videos.

**Review by aireenmei — COMMENTED** [review]
Currently mrope is missing from the sft pipeline and will be added in #4709, but #4709 focuses on image, I think some additional changes are required for video. How about we merge  #4709 first, then rebase and test again for video?

File: src/maxtext/input_pipeline/hf_data_processing.py [hunk 2]
```diff
@@ -42,6 +42,21 @@ def _get_pad_id(tokenizer):
   return pad_id
 
 
+def format_sharegpt(example):
```
[the remark this answers is not in the document]
Comment (hengtaoguo) [reply src/maxtext/input_pipeline/hf_data_processing.py:107]: Done, added strict list-type check for better control.

File: src/maxtext/input_pipeline/input_pipeline_utils.py [hunk 3]
```diff
@@ -90,19 +90,56 @@ def _process_string(string_tensor):
...
+    column,
+    image_placeholder,
+    model_name,
+    video_placeholder="<|video|>",
+    num_videos=0,
+    num_image_tokens=None,
+    num_video_tokens=None,
+):
   """reformat prompt for multimodal SFT"""
   if isinstance(example["images"], list):
```
[the remark this answers is not in the document]
Comment (hengtaoguo) [reply src/maxtext/input_pipeline/input_pipeline_utils.py]: Done and added the check. Also improve the function flow for better readability.

File: tools/data_generation/prepare_llava_video_178k.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+  parser = argparse.ArgumentParser(description="Prepare LLaVA-Video-178K dataset for MaxText.")
+  parser.add_argument("--output_dir", type=str, required=True, help="Directory to save the prepared dataset.")
+  parser.add_argument("--fold", type=str, default="0_30_s_academic_v0_1", help="Fold to download.")
+  parser.add_argument("--download_videos", action="store_true", help="Whether to download videos (large).")
+  parser.add_argument(
+      "--video_subset",
+      type=int,
+      default=None,
+      help="Number of video tars to download (1-8). If None, downloads all.",
+  )
```
[the remark this answers is not in the document]
Comment (hengtaoguo) [reply tools/data_generation/prepare_llava_video_178k.py:44]: This is working as intended. Video dataset is multi-gigabytes level and we don't expect different users run them multiple times.

File: src/maxtext/input_pipeline/input_pipeline_utils.py [hunk 5]
```diff
@@ -90,19 +90,56 @@ def _process_string(string_tensor):
...
 
 def reformat_response(example, column, model_name):
   """reformat response for multimodal SFT"""
-  example[column] = mm_processor.reformat_response(example[column][0], model_name)
+  if isinstance(example[column], (list, np.ndarray)):
+    response = example[column][0]
+  else:
+    response = example[column]
+  example[column] = mm_processor.reformat_response(response, model_name)
   return example
```
[the remark this answers is not in the document]
Comment (hengtaoguo) [reply src/maxtext/input_pipeline/input_pipeline_utils.py:130]: Fixed

File: src/maxtext/multimodal/processor_qwen3_omni.py (hunk 1 above)
Comment (lydhr) [reply src/maxtext/multimodal/processor_qwen3_omni.py:499]: If many videos fail silently, this might cause the model to train on irrelevant black frames, leading to quality regression.
Is it possible to skip or mask in input pipeline?

File: src/maxtext/multimodal/processor_qwen3_omni.py [hunk 6]
```diff
@@ -216,7 +218,7 @@ def downsample_video_mask_to_tokens(video_mask, config):
     raise ValueError(
         f"Video patch-mask length {patch_mask.shape[-1]} must be divisible by spatial merge area {merge_elements}."
     )
-  return patch_mask.reshape(patch_mask.shape[0], -1, merge_elements).max(axis=-1).astype(jnp.int32)
+  return patch_mask.reshape(patch_mask.shape[0], -1, merge_elements).max(axis=-1).astype(np.int32)
```
Comment (lydhr) [inline src/maxtext/multimodal/processor_qwen3_omni.py]: I wonder how much can this speed up cpu-bound ops. For the other jnp -> np in this file, are there any scenarios that might cause performance drops?

**Review by lydhr — APPROVED** [review]

File: src/maxtext/input_pipeline/input_pipeline_utils.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (hengtaoguo) [reply src/maxtext/input_pipeline/input_pipeline_utils.py:130]: I reversed this change to make the PR more simplistic without being overly defensive.

File: src/maxtext/multimodal/processor_qwen3_omni.py (hunk 6 above)
Comment (hengtaoguo) [reply src/maxtext/multimodal/processor_qwen3_omni.py]: Good catch, I did a small refactor to make sure:

1. Model-forward video-mask processing retains `jnp.int32`, since it executes on JAX arrays during the training process.
2. MRoPE data preprocessing below uses `np`, since it happens on host.

File: src/maxtext/multimodal/processor_qwen3_omni.py (hunk 1 above)
Comment (hengtaoguo) [reply src/maxtext/multimodal/processor_qwen3_omni.py:499]: Great suggestion! For my current test for 20 steps, I do see 2-3 videos were corrupted in the dataset and got switched to dummy videos. Let me do that as a follow up, namely skip the corrupted video during trainning.

**Review by abhinavclemson — APPROVED** [review]
